pycbc/types/timeseries.py

```diff
@@ -595,15 +595,34 @@ def gate(self, time, window=0.25, method='taper', copy=True,
             # Uses the hole-filling method of
             # https://arxiv.org/pdf/1908.05644.pdf
             from pycbc.strain.gate import gate_and_paint
+            from pycbc.waveform.utils import apply_fd_time_shift
             if invpsd is None:
                 # These are some bare minimum settings, normally you
                 # should probably provide a psd
                 invpsd = 1. / self.filter_psd(self.duration/32, self.delta_f, 0)
             lindex = int((time - window - self.start_time) / self.delta_t)
-            rindex = lindex + int(2 * window / self.delta_t)
+            rindex = int((time + window - self.start_time) / self.delta_t)
             lindex = lindex if lindex >= 0 else 0
             rindex = rindex if rindex <= len(self) else len(self)
-            return gate_and_paint(data, lindex, rindex, invpsd, copy=False)
+            rindex_time = data.sample_times[rindex]
+            offset = rindex_time - (time + window)
+            if offset == 0:
+                return gate_and_paint(data, lindex, rindex, invpsd, copy=False)
+            else:
+                # time shift such that gate end time lands on a specific data sample
+                fdata = data.to_frequencyseries()
+                fdata = apply_fd_time_shift(fdata, offset + fdata.epoch, copy=False)
+                # gate and paint in time domain
+                data = fdata.to_timeseries()
+                data = gate_and_paint(data, lindex, rindex, invpsd, copy=False)
+                # shift back to the original time
+                fdata = data.to_frequencyseries()
+                fdata = apply_fd_time_shift(fdata, -offset + fdata.epoch, copy=False)
+                tdata = fdata.to_timeseries()
+                # fill in projslc information
+                tdata.projslc = data.projslc
```

Comment:
This seems like a hack, would it not be better to have an argument that tells the function to return some additional things instead of just the data?

Reply:
@ahnitz This indeed looks like a hack to me as well! The two attributes, `tdata.projslc` and `tdata.proj` are so hidden. However, I'm imitating the return of the `gate_and_inpaint`, which was there originally. See here: https://github.com/gwastro/pycbc/blob/master/pycbc/strain/gate.py#L181

This returned format,  `tdata.projslc` and `tdata.proj`, is also later called by the gated_gaussian model, for example see this line: https://github.com/gwastro/pycbc/blob/master/pycbc/inference/models/gated_gaussian_noise.py#L565

I'd rather not touch any other files, so implement it in the current way to be consistent everywhere. Does this make sense? I hope I understand what you suggest correctly.

Reply:
@yi-fan-wang This can be removed. If you search for `current_proj` in `gated_gaussian_noise.py`, you'll see it isn't actually used anywhere; it was just for debugging purposes (e.g., see the comment [here](https://github.com/gwastro/pycbc/blob/master/pycbc/inference/models/gated_gaussian_noise.py#L49).) I would just comment out all the places `current_proj` is used in `gated_gaussian_noise.py`, then comment out those two lines of code in `timeseries.py` that adds the attribute. If @ahnitz prefers the lines be removed entirely you can do that too.

Reply:
done. I removed all the `projslc` and `proj`